waltz-client/src/main/java/com/wepay/waltz/client/TransactionContext.java

```diff
@@ -61,4 +68,10 @@ public void onCompletion(boolean result) {
     public void onException(Throwable ex) {
     }
 
+    /**
+     * A method that is call on failure of optimistic locking
+     */
+    public void onLockFailure() {
```

Comment:
Do you think it would be useful if application can decide whether or not to retry that log, probably by returning a boolean to this method?

Reply:
There can be a spurious lock failure (with a very small probability) due to a probabilistic nature of our lock table. I thought it is not safe to make retry decision in `onLockFailure()`.
Do you think it is useful to have an option to abort transaction even when it may be a false lock failure?

Reply:
Ah... I see, got it.

No, I think it is okay not to have that option.

(I was worried about usecases where synchronous api calls submit logs to Waltz n wait for those, but, in case of lock failure, app client can implement an expiry-like mechanism to return the api call rather than timeout waiting for the log to be retried n succeeded. When the api-call is retried, app client can check in its DB n determine whether or not to resubmit the log. If resubmitted, only one of the resubmitted or the being-retried logs will be successful due to optimistic locking.)